src/devices/cpu/upd177x/upd177x.cpp

```diff
@@ -0,0 +1,797 @@
+static constexpr u16 VECTOR_EXT   = 0x60;
+static constexpr u16 VECTOR_TIME  = 0x80;
+
+static constexpr u8 STACK_START = 0x20;
+
+
+upd177x_cpu_device::upd177x_cpu_device(const machine_config &mconfig, device_type type, const char *tag, device_t *owner, u32 clock)
+	: cpu_device(mconfig, type, tag, owner, clock)
+	, device_sound_interface(mconfig, *this)
+	, m_program_config("program", ENDIANNESS_BIG, 16, 16, -1, address_map_constructor(FUNC(upd177x_cpu_device::program_map), this))
+	, m_pa_out_cb(*this)
+	, m_pb_out_cb(*this)
+{ }
+
+
+upd1771c_cpu_device::upd1771c_cpu_device(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock)
+	: upd177x_cpu_device(mconfig, UPD1771C, tag, owner, clock)
+{ }
+
+
+device_memory_interface::space_config_vector upd177x_cpu_device::memory_space_config() const
+{
+	return space_config_vector
+	{
+		std::make_pair(AS_PROGRAM, &m_program_config)
+	};
+}
+
+
+void upd177x_cpu_device::program_map(address_map &map)
+{
+	map(0, 0x1ff).rom();
+}
+
+
+void upd177x_cpu_device::device_start()
+{
+	space(AS_PROGRAM).specific(m_program);
+
+	state_add(UPD177X_PC,        "PC",    m_pc);
+	state_add(STATE_GENPC,       "GENPC", m_pc).noshow();
+	state_add(STATE_GENPCBASE,   "CURPC", m_pc).noshow();
+	state_add(UPD177X_SP,        "SP",    m_sp);
+	state_add(UPD177X_A,         "A",     m_a);
+	state_add(UPD177X_A_SAVE,    "A'",    m_a_save);
+	state_add(UPD177X_H,         "H",     m_h);
+	state_add(UPD177X_N,         "N",     m_n);
+	state_add(UPD177X_TIMER,     "Timer", m_timer);
+	state_add(UPD177X_X,         "X",     m_x);
+	state_add(UPD177X_Y,         "Y",     m_y);
+	state_add(UPD177X_PA,        "PA",    m_pa);
+	state_add(UPD177X_PB,        "PB",    m_pb);
+	state_add(UPD177X_MD0,       "MD0",   m_md0);
+	state_add(UPD177X_MD1,       "MD1",   m_md1);
+	state_add(UPD177X_PNC1,      "PNC1",  m_pnc1);
+	state_add(UPD177X_PNC2,      "PNC2",  m_pnc2);
+	state_add(UPD177X_SKIP,      "SKIP",  m_skip);
+	state_add(UPD177X_SKIP_SAVE, "SKIP'", m_skip_save);
+
+	save_item(NAME(m_pc));
+	save_item(NAME(m_sp));
+	save_item(NAME(m_a));
+	save_item(NAME(m_a_save));
+	save_item(NAME(m_h));
+	save_item(NAME(m_n));
+	save_item(NAME(m_timer));
+	save_item(NAME(m_x));
+	save_item(NAME(m_y));
+	save_item(NAME(m_pa));
+	save_item(NAME(m_pb));
+	save_item(NAME(m_ram));
+	save_item(NAME(m_md0));
+	save_item(NAME(m_md1));
+	save_item(NAME(m_pnc1));
+	save_item(NAME(m_pnc2));
+	save_item(NAME(m_skip));
+	save_item(NAME(m_skip_save));
+	save_item(NAME(m_ts));
+	save_item(NAME(m_ns));
+	save_item(NAME(m_ss));
+	save_item(NAME(m_dac));
+	save_item(NAME(m_dac_sign));
+	save_item(NAME(m_tn_int));
+	save_item(NAME(m_counter));
+
+	set_icountptr(m_icount);
+
+	m_channel = stream_alloc(0, 1, clock() / 4);
+}
+
+
+void upd177x_cpu_device::device_reset()
+{
+	m_pc = 1; // Yes, this is odd, but otherwise there are sync issues between writing and clearing PA in scv kungfurd when playing back adpcm.
```

Comment:
That's an elegant hack.

FWIW, on silicon, an external write (from the main CPU) takes priority over 'OUT PA'. And because the 'OUT PA' completes while the external write of 0x1F is still in progress, the external write wins.

Reply:
Thanks for that background information. I already thought it should be something to do with low level timing between the chips.
I have updated the comment on that hack.